docs/modelplugin.md

```diff
@@ -0,0 +1,175 @@
+# Extending onos-config with Model Plugins
+
+onos-config is an extensible configuration management system, that allows the
+configuration of many different types and versions of devices to be managed
+concurrently.
+
+Information models in YANG format (RFC 6020) can be used to accurately define
+the configuration and state objects and attributes of a device. In practice a
+device's object model usually comprises of a number of YANG files including
+augments and deviations, and must be considered as a combined unit.
+
+In **onos-config** a set of these combined YANG files defining a particular
+version of a device type is known as a **model**.
+
+Over its lifecycle onos-config will have to deal with many different models as
+its scope is expanded and as devices go through new release cycles. To allow
+this models are loadable dynamically as plugins in the form of Linux or Mac
+shared object libraries **(\*.so)** using the YGOT library and are know as
+**Model Plugins**.
+
+The diagram shows the connection between the Model Plugin and the configuration
+store - linked by Device Type and Version. Effectively the primary key
+of the Model Registry is the Model Name and Version, whereas the primary key of
+the Configuration is the Device Name and Version. 
+
+![onos-config internals](images/onos-config-internals.png)
+
+## Structure of a Model Plugin
+A Model Plugin is mainly generated by the generator command from the YGOT
+project, and a wrapper **modelmain.go** implementing the ModelPlugin interface.
+They are compiled together with the **go build** command using the
+**-buildmode=plugin** option. 
+
+Many examples of Model Plugins are in the [../modelplugins](../modelplugins)
+folder of this project, and an example script
+[ModelGenerator.sh](../modelplugins/ModelGenerator.sh) is available for creating
```

Comment:
can you show an example of how to run both the script and the go build ?

Reply:
done